Rigid body plant mux + PidControlledSystem without plant ownership

sammy-tri opened the pull request:
This PR combines some prerequisites to my combined LCM controlled iiwa+Schunk work, namely having multiple LCM command sources and controllers operating on the same RigidBodyPlant.

It's possible RigidBodyPlantMux is generic enough (and good enough in this implementation) to move outside of the kuka_iiwa_arm example.  It's also possible the idea is so ill-conceived that it shouldn't be merged at all.  Feedback appreciated.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4475)
<!-- Reviewable:end -->

sammy-tri commented:
+@liangfok for feature review based on earlier discussion from Slack which indicates this could be relevant to his interests (and aleady being familiar with some of the affected code)

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-:-KYoHcCTkVde8GUtADi2:b-ej2tbd)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
I just gave it a quick pass. Will swing by again later today. PTAL thanks!

---

Reviewed 2 of 7 files at r1.
Review status: 2 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h, line 31 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYrwwk2MV_Ipw2XPkfM:-KYrwwk3cNPe6Vi4skh8:b-1tmqd) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h#L31)):*
> ```objc
>   /// Creates a new vector values output port where the values of the
>   /// port correspond to @p position_names followed by @p
>   /// velocity_names.
> ```

Could you clarify what is a position_name and a velocity_name? 

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYrx82J82OxsF-YSJD3:-KYrx82K01pi-baiBIOF:b-tv479o) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h#L38)):*
> ```objc
>   /// Using @p builder, connect the ports previously created with
>   /// AddInput and AddOutput to @p plant.  It is an error to call
>   /// AddInput or AddOutput after ConnectPlant.
> ```

Maybe this condition can be enforced with an internal bool flag? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-:-KYs2xc6GnuudKFDvvmo:bw5hnoi)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 2 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h, line 31 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYrwwk2MV_Ipw2XPkfM:-KYtJAbmQGOy_WGQ5SNw:bauul8l) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h#L31)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

Could you clarify what is a position_name and a velocity_name? 
</blockquote></details>

Added some (hopefully) clarifying text.

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h, line 38 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYrx82J82OxsF-YSJD3:-KYtJGh4Efs2BWh5bWkR:bcum676) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h#L38)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

Maybe this condition can be enforced with an internal bool flag? 
</blockquote></details>

Added additional DRAKE_DEMAND in the implementation code (the input_mux_ and output_mux_ members will be invalidated when they're added to the system).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4475)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I left some high-level comments and will perform a full feature review after they are addressed.

---

Reviewed 2 of 7 files at r1.
Review status: 1 of 7 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.cc, line 18 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYsebR_QH9FVnzETHGu:-KYsebRaA0_3Pkbe5dcm:b-wsbyqx) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.cc#L18)):*
This assumes that each actuator only contains one output. Please add the following `TODO` comment:

> TODO(liangfok) Generalize to support multi-DOF actuators once #4153 is resolved.

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.cc, line 18 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYt8YMzQqXggqCIzNFT:-KYt8YMzQqXggqCIzNFU:b-9t11zn) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.cc#L18)):*
> ```cpp
>   explicit InputMux(const RigidBodyTree<T>& tree) {
>     // Input ports will be declared when requested.
>     this->DeclareOutputPort(systems::kVectorValued, tree.get_num_actuators());
> ```

Per F2F discussion, please generalize this class to support `RigidBodyTree` objects that contain more than one instance of the robot. To do this, the actuator name needs to be used in conjunction with a `model_instance_id`. The `model_instance_id` is returned by the parser method that adds the model instance to the `RigidBodyTree`.

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.cc, line 20 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYseH8tqtgr7nMHIm0D:-KYseH8tqtgr7nMHIm0E:b-q3op77) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.cc#L20)):*
> ```cpp
>     this->DeclareOutputPort(systems::kVectorValued, tree.get_num_actuators());
> 
>     for (int i = 0; i < tree.get_num_actuators(); i++) {
> ```

BTW: nit: prefer `++i` for consistency.

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYsmp-gr_rHkrn4T9O0:-KYsmp-gr_rHkrn4T9O1:bc71d9z) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h#L17)):*
> ```objc
> /// Creates multiplexers to wrap the input/output ports of a
> /// RigidBodyPlant to allow for multiple controllers to operate on
> /// different parts of the plant.
> ```

BTW, consider updating this documentation to describe the number and nature of input and output ports exported by this system.

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYseKuPJCcPFWMs2XGE:-KYseKuPJCcPFWMs2XGF:b53hu79) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h#L21)):*
Please add documentation to this method:

> A constructor that initializes this system's single output port, which contains the outputs of the tree's actuators.

---

*[drake/systems/controllers/pid_controlled_system.h, line 143 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYt8yAl_0prr1s3nhA3:-KYt8yAl_0prr1s3nhA4:b-welbi0) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/systems/controllers/pid_controlled_system.h#L143)):*
> ```objc
>   }
> 
>   /// Connect a PidController to an existing plant and DiagramBuilder,
> ```

Per style guide, "Connect" should be "Connects".

---

*[drake/systems/controllers/pid_controlled_system.h, line 143 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYtKHjyRWogbIVNA3AH:-KYtKHjyRWogbIVNA3AI:b-stjfqv) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/systems/controllers/pid_controlled_system.h#L143)):*
> ```objc
>   }
> 
>   /// Connect a PidController to an existing plant and DiagramBuilder,
> ```

Perhaps clarify that this method "Creates and connects...".

---

*[drake/systems/controllers/pid_controlled_system.h, line 144 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYtA9dRa5LBKMmSLPgw:-KYtA9dRa5LBKMmSLPgx:b-4md9l5) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/systems/controllers/pid_controlled_system.h#L144)):*
> ```objc
> 
>   /// Connect a PidController to an existing plant and DiagramBuilder,
>   /// adding additional systems as needed.  Returns a pair of port
> ```

Please document what types of additional systems could potentially be added and under what circumstances.

---

*[drake/systems/controllers/pid_controlled_system.h, line 150 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYtExW8vjdLIHYmb48j:-KYtExW8vjdLIHYmb48k:b-c9cn0s) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/systems/controllers/pid_controlled_system.h#L150)):*
> ```objc
>   static std::pair<const SystemPortDescriptor<T>,
>                    const SystemPortDescriptor<T>> ConnectController(
>                        const SystemPortDescriptor<T>& plant_input,
> ```

Can the plant's input and output ports be obtained from the `DiagramBuilder` using `GetMutableSystems()` assuming there is only one plant within the `DiagramBuilder`? Do we not want to make this assumption?

---

*[drake/systems/controllers/pid_controlled_system.h, line 156 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYtFj5B8KUrfV50RiuW:-KYtFj5B8KUrfV50RiuX:b-rz684e) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/systems/controllers/pid_controlled_system.h#L156)):*
> ```objc
>                        const VectorX<T>& Kd,
>                        DiagramBuilder<T>* builder,
>                        PidController<T>** controller);
> ```

Instead of making this an output parameter, how about modifying this method to return a `unique_ptr<PidController<T>>`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-:-KYt8rdSu8hW75g1ml6Q:b-hv190h)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 1 of 7 files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h, line 25 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYtKzfuzyqFV8onuZ62:-KYtKzfuzyqFV8onuZ63:bzc24bl) ([raw file](https://github.com/robotlocomotion/drake/blob/862a7e7474f6f31e19b639a1337bb364f0991dfa/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h#L25)):*
> ```objc
> 
>   /// Creates a new vector valued input port where the values of the
>   /// port correspond to the actuators from the rigid boty tree
> ```

Typo: "boty".

Consider replacing "rigid body tree" with "RigidBodyTree" show it shows up as a link in the generated Doxygen website.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4475)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 1 of 7 files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.cc, line 18 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYt8YMzQqXggqCIzNFT:-KYuCRwUdjAga-2A3XwH:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.cc#L18)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Per F2F discussion, please generalize this class to support `RigidBodyTree` objects that contain more than one instance of the robot. To do this, the actuator name needs to be used in conjunction with a `model_instance_id`. The `model_instance_id` is returned by the parser method that adds the model instance to the `RigidBodyTree`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4475)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 1 of 7 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.cc, line 20 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYseH8tqtgr7nMHIm0D:-KYuQ5fY_a7ISwW6Ft_s:bvoqcmk) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.cc#L20)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW: nit: prefer `++i` for consistency.
</blockquote></details>

Consistency with what?  This file is internally consistent, and I don't recall the style guide mandating pre-increment for POD types?   That being said, I should probably just get in the habit eventually.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4475)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 1 of 7 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.cc, line 18 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYsebR_QH9FVnzETHGu:-KYuSTYwT-Pob7OZRtKk:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.cc#L18)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This assumes that each actuator only contains one output. Please add the following `TODO` comment:

> TODO(liangfok) Generalize to support multi-DOF actuators once #4153 is resolved.
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYsmp-gr_rHkrn4T9O0:-KYuQkutQA1xO__yEY13:b-ewet9z) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h#L17)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider updating this documentation to describe the number and nature of input and output ports exported by this system.
</blockquote></details>

It's not a system, and doesn't directly have any ports due to this.  Any suggestions on how to clarify the documentation to indicate that fact?

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h, line 21 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYseKuPJCcPFWMs2XGE:-KYuQurwhXf7h1qi5LoC:b-3ctfix) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h#L21)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Please add documentation to this method:

> A constructor that initializes this system's single output port, which contains the outputs of the tree's actuators.
</blockquote></details>

See above, this isn't actually a system.  I guess it does create some systems which are transferred into the DiagramBuilder in ConnectPlant...

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h, line 25 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYtKzfuzyqFV8onuZ62:-KYuR5ZJvVZmcigLX2mn:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/862a7e7474f6f31e19b639a1337bb364f0991dfa/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h#L25)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Typo: "boty".

Consider replacing "rigid body tree" with "RigidBodyTree" show it shows up as a link in the generated Doxygen website.
</blockquote></details>

Done.

---

*[drake/systems/controllers/pid_controlled_system.h, line 143 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYt8yAl_0prr1s3nhA3:-KYuRC-nmc_vAQqB4eKE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/systems/controllers/pid_controlled_system.h#L143)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Per style guide, "Connect" should be "Connects".
</blockquote></details>

Done.

---

*[drake/systems/controllers/pid_controlled_system.h, line 143 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYtKHjyRWogbIVNA3AH:-KYuRMVC6XCv2gKYlrMu:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/systems/controllers/pid_controlled_system.h#L143)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Perhaps clarify that this method "Creates and connects...".
</blockquote></details>

Done.

---

*[drake/systems/controllers/pid_controlled_system.h, line 144 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYtA9dRa5LBKMmSLPgw:-KYuRaW1nLfy3YTlyotF:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/systems/controllers/pid_controlled_system.h#L144)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Please document what types of additional systems could potentially be added and under what circumstances.
</blockquote></details>

Done.

---

*[drake/systems/controllers/pid_controlled_system.h, line 150 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYtExW8vjdLIHYmb48j:-KYuRbsbKEZAaVKlOnpI:bqfghaf) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/systems/controllers/pid_controlled_system.h#L150)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Can the plant's input and output ports be obtained from the `DiagramBuilder` using `GetMutableSystems()` assuming there is only one plant within the `DiagramBuilder`? Do we not want to make this assumption?
</blockquote></details>

I would expect that the plant's input and export port descriptors would normally come from the plant itself and be handled by the caller before we get here.  I don't think GetMutableSystems should be necessary.

---

*[drake/systems/controllers/pid_controlled_system.h, line 156 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYtFj5B8KUrfV50RiuW:-KYuS4CFbf5-9lvS_ars:b-ks93zi) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/systems/controllers/pid_controlled_system.h#L156)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Instead of making this an output parameter, how about modifying this method to return a `unique_ptr<PidController<T>>`?
</blockquote></details>

In addition to the std::pair it's already returning?  I guess I could return a std::tuple with the unique_ptr in it too perhaps.  Actually, I should look and see how important returning the controller even is now that the state defaults to zero.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4475)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I left some responses; still haven't done a full feature review.

---

Review status: 1 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.cc, line 20 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYseH8tqtgr7nMHIm0D:-KYuzyUFb6KVdAbHzdMh:b8660uv) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.cc#L20)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Consistency with what?  This file is internally consistent, and I don't recall the style guide mandating pre-increment for POD types?   That being said, I should probably just get in the habit eventually.
</blockquote></details>

Sorry, I meant to say "consistent with when pre-vs-post increment actually matters.

Here is the style guide rule:

https://google.github.io/styleguide/cppguide.html#Preincrement_and_Predecrement

It only establishes a pre-increment bias for "iterators and other template types". However, to be consistent with iterators and other template types, we might as well use pre-increment for scalar types as well.

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYsmp-gr_rHkrn4T9O0:-KYvHoxVRjLISm8BewWF:b-k2xod9) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h#L17)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

It's not a system, and doesn't directly have any ports due to this.  Any suggestions on how to clarify the documentation to indicate that fact?
</blockquote></details>

Oh, interesting. Have you considered making `RigidBodyPlantMux` a `Diagram` that instantiates and wires up Mux / DeMux systems to a `RigidbodyPlant`?

Alternatively, consider renaming this class `ControlledRigidBodyPlantBuilder` to reduce the likelihood that a reader would confuse this class with a `System`.

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h, line 35 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYv0eKCaPxHYPmeQgpv:-KYv0eKCaPxHYPmeQgpw:bkfas6k) ([raw file](https://github.com/robotlocomotion/drake/blob/ba086a65e47187361b9e75e18ef5d1a2e51d19cd/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h#L35)):*
> ```objc
>       const std::vector<std::pair<std::string, int>>& actuators);
> 
>   /// Creates a new vector values output port where the values of the
> ```

BTW, I believe "vector values" should be "vector value" since "vector" is singular.

---

*[drake/systems/controllers/pid_controlled_system.h, line 156 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYtFj5B8KUrfV50RiuW:-KYvHJs83abfNA0byfA4:b-2idv78) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/systems/controllers/pid_controlled_system.h#L156)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

In addition to the std::pair it's already returning?  I guess I could return a std::tuple with the unique_ptr in it too perhaps.  Actually, I should look and see how important returning the controller even is now that the state defaults to zero.
</blockquote></details>

Oh, maybe I'm mistaken. Are the returned port descriptors _not_ ports exported by `@p controller`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-:-KYvIz9posndyezlxA_q:b-x5iat1)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 1 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.cc, line 20 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYseH8tqtgr7nMHIm0D:-KYxs9PXI3Nwg40Rvp0O:bhp9nqn) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.cc#L20)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Sorry, I meant to say "consistent with when pre-vs-post increment actually matters.

Here is the style guide rule:

https://google.github.io/styleguide/cppguide.html#Preincrement_and_Predecrement

It only establishes a pre-increment bias for "iterators and other template types". However, to be consistent with iterators and other template types, we might as well use pre-increment for scalar types as well.
</blockquote></details>

It explicitly allows either form, and we don't have an override in our style guide.  That being said, I caved anyway and changed this file.

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYsmp-gr_rHkrn4T9O0:-KYxsgU51BGSSvjxOqwe:b68qhtv) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h#L17)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Oh, interesting. Have you considered making `RigidBodyPlantMux` a `Diagram` that instantiates and wires up Mux / DeMux systems to a `RigidbodyPlant`?

Alternatively, consider renaming this class `ControlledRigidBodyPlantBuilder` to reduce the likelihood that a reader would confuse this class with a `System`.
</blockquote></details>

I thought about making it a diagram vs. allowing it's components to incorporate into another diagram.  I didn't have a strong preference either way, and I found this fairly easy to incorporate into the code I was writing it for so I went with this style.

It doesn't actually create the plant directly, so I'm not sure "Builder" is the best term to use.  (I don't love "Mux" either.  I wonder if I'll eventually get defected for using an abbreviation?)

---

*[drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h, line 35 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYv0eKCaPxHYPmeQgpv:-KYxtAsDG4yNXv9lBZoG:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ba086a65e47187361b9e75e18ef5d1a2e51d19cd/drake/examples/kuka_iiwa_arm/rigid_body_plant_mux.h#L35)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, I believe "vector values" should be "vector value" since "vector" is singular.
</blockquote></details>

Done.

---

*[drake/systems/controllers/pid_controlled_system.h, line 156 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-KYtFj5B8KUrfV50RiuW:-KYxtNw7I0EEu7YCM56H:bibe3v3) ([raw file](https://github.com/robotlocomotion/drake/blob/540acf0565c64f62d4e3eeea04af7f485432a2ad/drake/systems/controllers/pid_controlled_system.h#L156)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Oh, maybe I'm mistaken. Are the returned port descriptors _not_ ports exported by `@p controller`?
</blockquote></details>

The first port is one of the ports on the Adder which combines the feed forward control input with the controller's output, and the second is the input to the "error inverter" (which then takes a somewhat tortured path to get to the actual PidController).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4475)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
+(status: do not merge) blocking merge until we decide what's up with #4500 (Generalize RigidBodyPlant's Input and Output Ports To be Model Instance Specific) 

---

Review status: 1 of 7 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4475#-:-KYya2GVa_Oq1SDHGZSF:bwepuk4)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 7 files at r1, 3 of 5 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4475)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Closing in favor of the approach in #4506
